### DISCUSSION TITLE
Reliable Cheat Detection Method Using Statistics

### DISCUSSION THREAD

**Discussion by dananman** [ppy/osu#23194] [Ideas] [answered]
- - - -
- - - -
osu! is a popular, simple, skill based game which makes it a target for cheaters and cheat developers.
With the increase in the player-base there is also an increase in exceptionally fast improving players as well as in smart cheaters which makes it harder with time to differentiate between the two.
- - - -
But osu!'s popularity, simplicity and a focus on skill makes it easy to gather accurate, detailed and usefully relevant statistics regarding player skill and improvement overtime. I think there's an opportunity to gather the relevant statistics to create reliable models of natural player improvement in many skills in the game which will make it much easier to single out outliers early on and demand additional information (e.g. live play) to decide if players that stand on the intersection between fast improvers and smart cheaters are legit or not.
It will also give much greater confidence in and ability to auto restrict players who show statistically impossible improvement and skill.
- - - -
- - - -
#### Player consent
There shouldn't be any concerns from the players regarding the data collection as the data is not personal nor will be used for things like personalized ads. Furthermore, there is already some data collection of this sort in osu! that players know and even like - play count, average accuracy, rank history etc.  Players probably will be glad to help osu! stuff better detect cheaters and maybe even will find their personal data useful for themselves if they'll be able to see it as it will give better insight into their skill and improvement.

#### What data will be tracked and collected
- Devices used (keyboard/keypad/feet, mouse/tablet/touch/anything else)
- Hit timing accuracy for each object
- Hit precision for each object
- AR, OD, CS of each object
- Distance from each object to the previous object
- Time past between each object and the previous object
- When special mods are used like HD, FL, SO
- Anything else useful I haven't thought of
- Time playing and time resting/not playing

#### How to organize the data
There are three distinct useful groups to initially organize the data:

**Player environment (A):**

1. Devices used (keyboard/keypad/feet, mouse/tablet/touch/anything else)
2. AR, OD, CS of each object
3. Distance from each object to the previous object
4. Time past between each object and the previous object
5. When special mods are used like HD, FL, SO

**Player performance (B):**

1. Hit timing accuracy for each object
2. Hit precision for each object

**play frequency (C):**

1. Time playing and time resting/not playing

From groups (A) & (B) we need to create ten chart categories by charting each item from (A) with an item from (B) for each note a player encounters within each single play that has some minimum amount of notes (to not account for mindblock retry spam). We calculate the average values for each chart and then accounting for (C) to track improvement between plays while accounting for rest between each play and session.
This will give us the data on each player's improvement for each category.

If the data will be collected for long enough, and there will be a large amount players in the data, it would be possible to create models that predict player improvement given past improvement.
When a player deviates from the model they previously fit by improving too fast in some or all categories by a statistically significant amount, we can then single them out and decide what to do.

- - - -

This idea is just a rough blueprint that gives the general idea of how it can be done. There probably be much more nuance and statistical work to be done, as well as accounting for more factors like reading difficulty, pattern angels and so on, but I believe that it is worth considering as it may be very effective at detecting cheating early on, preventing cheaters from getting to high ranks and generally discouraging people from trying to cheat.

- - - -

### Other Possible Uses For The Data

Other than cheat detection, the data can be used to recognize multi-accounting and can provide users which graphs and charts of their improvement (maybe part of it will be accessible only for supporters).
The data can possibly also help to better the PP system by showing what skills are faster or slower learned, more rare, and which patterns and types of maps are harder and which are easier.

**Comment by ItsShamed** [comment, marked as answer]
Just so you know:
- Internal work has already started regarding the AC
- osu! already stores replays or that is already planned (https://github.com/ppy/osu-server-spectator/pull/155)

Also, everything you said you want to collect is already available online for the most part, and we technically have more than enough data needed to “create a model” *(hell, people even managed to do it for player movement, see https://github.com/minetoblend/osucad-cursor-predict and [this tweet](https://twitter.com/minetoblend/status/1587078396558954497))*

So I think you are writing a lot of text just to suggest to use Machine Learning for cheat detection, which has already been done on other games and is a direction that could be taken, but this needs consideration.

**Reply by dananman** [reply]
oh I see, I hope it'll be implemented!
